{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fa1de821",
   "metadata": {},
   "source": [
    "## Course 1 - Finding Hidden Messages in DNA  \n",
    "\n",
    "#### The course deals with following topics:\n",
    "  1.1 Finding/counting patterns in DNA string, accounting for limited mismatches/reverse complement  \n",
    "  1.2 Finding a shared motif from a set of DNA strings, allowing for mismatches  \n",
    "  \t- Greedy search, Monte Carlo and Gibbs Sampling  \n",
    "  1.3. Final project: Finding transcription factor binding motif for MTB genes upregulated in hypoxia."
   ]
  },
  {
   "cell_type": "raw",
   "id": "14a88227",
   "metadata": {},
   "source": [
    "## Course 1 - Finding Hidden Messages in DNA  \n",
    "\n",
    "#### The course deals with following topics:\n",
    "  1.1 Finding/counting patterns in DNA string, accounting for limited mismatches/reverse complement  \n",
    "  1.2 Finding a shared motif from a set of DNA strings, allowing for mismatches  \n",
    "  \t- Greedy search, Monte Carlo and Gibbs Sampling  \n",
    "  1.3. Final project: Finding transcription factor binding motif for MTB genes upregulated in hypoxia."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "820caa8e",
   "metadata": {},
   "source": [
    "### Frequent Word Problem\n",
    "Solve the Frequent Words Problem.\n",
    "  * Input: A string Text and an integer k.\n",
    "  * Output: All most frequent k-mers in Text."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "1f0dc9dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def FrequentWords(Text, k):\n",
    "    PatternsInText = []\n",
    "    NumPatterns = []\n",
    "    for i in range(len(Text)-k+1):\n",
    "        Found = 0\n",
    "        if Text[i:i+k] in PatternsInText:\n",
    "            j = PatternsInText.index(Text[i:i+k])\n",
    "            NumPatterns[j] += 1\n",
    "            Found = 1\n",
    "        if Found == 0:\n",
    "            PatternsInText.append(Text[i:i+k])\n",
    "            NumPatterns.append(1)\n",
    "        maxpos = [i for i in range(len(NumPatterns)) if NumPatterns[i]==max(NumPatterns)]\n",
    "        FrequentKmer = [PatternsInText[i] for i in maxpos]\n",
    "    return FrequentKmer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "9e3c2966",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Text = TTGCGCCTAAATGATCAATGATCTTGCGCCTAAATGATCGACTGTAGTTGCGCCTACGGAGATTTGCGCCTAGACTGTAGCGGAGATGACTGTAGACTCATTGAGACTGTAGACTCATTGATTGCGCCTACGGAGATGACTGTAGCGGAGATTTGCGCCTATTGCGCCTAAATGATCAATGATCGACTGTAGAATGATCCGGAGATCGGAGATCGGAGATGACTGTAGCGGAGATACTCATTGAAATGATCACTCATTGATTGCGCCTACGGAGATGACTGTAGCGGAGATACTCATTGACGGAGATTTGCGCCTAACTCATTGACGGAGATGACTGTAGGACTGTAGGACTGTAGTTGCGCCTAAATGATCGACTGTAGTTGCGCCTAACTCATTGAGACTGTAGTTGCGCCTAACTCATTGAGACTGTAGGACTGTAGACTCATTGAACTCATTGAAATGATCAATGATCCGGAGATAATGATCTTGCGCCTAAATGATCAATGATCCGGAGATCGGAGATAATGATCCGGAGATCGGAGATAATGATCGACTGTAGTTGCGCCTACGGAGATAATGATCACTCATTGACGGAGATAATGATCCGGAGATTTGCGCCTAGACTGTAGACTCATTGAGACTGTAGGACTGTAGTTGCGCCTAGACTGTAGTTGCGCCTATTGCGCCTAACTCATTGAACTCATTGATTGCGCCTACGGAGATGACTGTAGCGGAGATCGGAGATCGGAGATCGGAGATTTGCGCCTATTGCGCCTATTGCGCCTACGGAGATCGGAGATTTGCGCCTATTGCGCCTAGACTGTAGGACTGTAGACTCATTGAACTCATTGATTGCGCCTACGGAGAT\n",
      "\n",
      "Pattern = 14\n",
      "\n",
      "Words = CGGAGATCGGAGAT\n"
     ]
    }
   ],
   "source": [
    "fnm = 'Input_FrequentWords.txt'\n",
    "with open('Course1_Data/'+fnm,'r',encoding='utf-8') as file:\n",
    "    Text = file.readline().strip()\n",
    "    k = int(file.readline().strip())\n",
    "FrequentKmer = FrequentWords(Text, k)\n",
    "\n",
    "print(f'Text = {Text}\\n')\n",
    "print(f'Pattern = {k}\\n')\n",
    "print('Words = '+' '.join(FrequentKmer))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f524e3a5",
   "metadata": {},
   "source": [
    "### Pattern Matching Problem.\n",
    "\n",
    "  * Input: Two strings, Pattern and Genome.\n",
    "  * Output: A collection of space-separated integers specifying all starting positions where Pattern appears as a substring of Genome.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "d3106d4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def PatternMatching(Pattern, Genome):\n",
    "    position = []\n",
    "    for j in range(len(Genome)-len(Pattern)+1):\n",
    "        if Genome[j:j+len(Pattern)] == Pattern:\n",
    "            position.append(j)\n",
    "    return positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "76cd3d96",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pattern = AGACTCCAG\n",
      "\n",
      "Genome = ACCAGACTCCTACAGACTCCGCATAGACTCCGGAGACTCCTAGACTCCAGACTCCAGACTCCGGGAGACTCCCTAGAGACTCCGCAAGACTCCAGACTCCAGAGACTCCAAAGACTCCAGACTCCGAGACTCCAGACTCCTAGACTCCCGAAGACTCCGAGACTCCCGAGACTCCGGATGTGAGACTCCAGACTCCTAAGACTCCAGACTCCATAAGACTCCAGACTCCGAGACTCCCAAGACTCCAGAGCTGCAGACTCCTATAGACTCCAGAGCCAGACTCCTAGACTCCAGACTCCTTGAGAGACTCCCCGTTCGAGACTCCAGACTCCAATAGAAACAGACTCCGTGAGCAGACTCCAAGTACAGACTCCCCGGCAGACTCCAGACTCCAGACTCCTAGACTCCAGACTCCAGACTCCTAGACTCCAAGACTCCTGTCAGACTCCCGCGGGAGACTCCGTAAGTCTCAGTCTTAGACTCCAAGACTCCAAGTTTATACATAGACTCCCAGACTCCAGACTCCCAGACTCCAGTTAGACTCCTAGACTCCTAAGACTCCTAGACTCCGATGCAGACTCCATATAGACTCCGCAGACTCCAGGGGGTCAGTTTCTTGTAGACTCCTAGGGAAGACTCCAGACTCCTTGAGACTCCTAAGACTCCGCGAGACTCCCAGACTCCAGACTCCAGACTCCAGACTCCACCGAGACTCCAGACTCCAGACTCCAGACTCCCTCACAGACTCCAGACTCCCCTAGACTCCAGAGACTCCATAGACTCCCAAGACTCCAGACTCCGCAGACTCCTGTAGACTCCGAGACTCCAGAGACTCCCGAGACTCCGTCAGACTCCAGACTCCAGACTCCTAGACTCCAGACTCCAGACTCCAGACTCCCAAGACTCCCATTAGACTCCTGAGACTCCCTAGACTCCGAGACTCCAGACTCCTGTCAGACTCCTCGGCAAGACTCCGTCAGACTCCAAGACTCCAGACTCCTAGACTCCGACTTCAGACTCCGAAGACTCCACAGTAGACTCCGGGCCAAGACTCCAGACTCCAAGACTCCGAGACTCCGAGACTCCAGACTCCGAGACTCCTCAAGACTCCTCGAGACTCCGCTGCTTCCCAGACTCCTGTAAGACTCCTAGACTCCAGACTCCTGAGACTCCCAGACTCCAGACTCCAGACTCCGAGACTCCTAAGACTCCGTTTAGACTCCAGACTCCAAGACTCCGAGACTCCCAGACTCCTGATAGACTCCTCTCAGACTCCACAAACAAGACTCCGCAGACTCCTCGGTAGACTCCAGACTCCAAGACTCCAGACTCCTACAGACTCCAGACTCCATAGGGGGAGACTCCGGCTTAGACTCCAGACTCCTGAGACTCCAGACTCCAGACTCCAGACTCCAGACTCCAGACTCCCAGACTCCAGACTCCCGGTCGGAGACTCCGATGGAGACTCCACAGACTCCAGACTCCTAAGACTCCTTACAAGACTCCAAGACTCCGACCAGACTCCAGACTCCTCGACAGACTCCAAGACTCCAGATACAGACTCCTTCAGACTCCATTGACCAGACTCCATAAGACTCCTAGACTCCAAGACTCCAAGACTCCTAGACTCCCCTGAATAGACTCCGAGACTCCCAGACTCCAGACTCCCACACAGACTCCGAGACTCCCGGACCAGACTCCAGACTCCGCATAACTAGACTCCAGCTCTAGACTCCGAAGACTCCAGACTCCGAGACTCCAGACTCCCAGACTCCAGACTCCTAGTAGACTCCGAGACTCCGTAGACTCCGAGACTCCAGACTCCGACTTGGAAAGACTCCAGACTCCCATGTAGACTCCGAGCGGAGACTCCAAGACTCCAGACTCCGTAGACTCCGGAGACTCCAAGAGACTCCAGACTCCAGAGACTCCAAGACTCCCAATGAGACTCCCAGACTCCAGACTCCTCAGACTCCCTCTGGACATAGACTCCACCAGACTCCTGAGACTCCAGACTCCAGACTCCACAGACTCCCAAGACTCCAAAGACTCCTCAGACTCCAGGGGGAAGACTCCGTAACAGACTCCAGACTCCAATTCTCAGACTCCAGACTCCGGTCAAGACTCCCAAGACTCCGGCAGACTCCAAGACTCCTAGACTCCAGACTCCTGAGACTCCAGACTCCGGGAGACTCCCCTTAGACTCCAGACAGACTCCGGAGACTCCCAGACTCCTTAGAGACTCCTAGACTCCTGTAGTCTTAGCCAAGACTCCAAGACTCCATGCAGCAAGACTCCCAGACTCCCGTTAAGACTCCAAGCAACAGACTCCCCCAGACTCCGATAGACTCCAAGACTCCAGACTCCGGTGTCCAGACTCCAGACTCCAGACTCCTCAGACTCCGAAGACTCCGACAGGGTAAAGACTCCATTTTACCAGACTCCTGCACAGACTCCGCCGTACAATAAGACTCCCCAGACTCCAGACTCCAGACTCCAGACTCCAAGACTCCAGACTCCTGAGACTCCACAGACTCCCAGACTCCCTCGACCGTATGGAAAAGACTCCTAGACTCCATAGACTCCAGACTCCCAAGACTCCAGACTCCTAGAGACTCCGAGACTCCAATTTTCAGACTCCTGACACAAGACTCCAATTAGACTCCAGACTCCAAAGACTCCTTATTGCTAGACTCCCTTGAGACTCCACCAGACTCCGCAAAGACTCCAGACTCCACAGACTCCAGACTCCTACAGCTAGCAGACTCCAGACTCCTTATTAGACTCCAGACTCCAATCGCAGACTCCCGCCCAAGACTCCAGACTCCTAGACTCCGCAGACTCCTTAGACTCCAGAAGACTCCAAGACTCCTACAGACTCCGAGACTCCCACTAGGAGACTCCTAAGACTCCAGACTCCACAGACTCCAGACTCCAGACTCCAGACTCCTTGAGACTCCACGAGACTCCGAGACTCCTAAAGACTCCAGACTCCAGACTCCGTCCAGACTCCGTAGACTCCCAGACTCCTAGACTCCAGACTCCGGAGGACTCAAGACTCCTACAGACTCCCAGACTCCCTCCGAGACTCCAGACTCCCAGACTCCACAGACTCCAGACTCCAAGACTCCGTAGACTCCAGACTCCAGACTCCACCCACTAAGACTCCAAGACTCCGTTAGACTCCCAGAAGACTCCCCTTAGACTCCCGTATAGACTCCCAGACTCCCAGACTCCAGACTCCGGTCGTAGGCATAGACTCCAGACTCCAGACTCCGTCTGTCCGTCGAGACTCCGACAGACTCCAGACTCCGAGACTCCTCACTTCACCGCAGACTCCCTGCAAGACTCCTAGACTCCGAGACTCCCTTAGAGAGACTCCGTCAGACTCCACGAGACTCCAGTAAGACTCCTGATTGAGACTCCATAGACTCCAGACTCCCCAGACTCCAGACTCCTAGACTCCAGACTCCAAGACTCCAGACTCCAGACTCCCAGACTCCGAATGCAGACTCCGGGCAAGACTCCAGACTCCCTAGACTCCGGGTGGAGACTCCGGAGAGACTCCAGACTCCAGACTCCAGACTCCGAGACTCCAAGACTCCAGACTCCCAGACTCCAGACTCCGACGCAGACTCCAGACTCCAGACTCCTAGACTCCAAAGACTCCTAACTCGGAGACTCCCTAGACTCCGAAAAGACTCCGAGACTCCTAGGAGACTCCTAGACTCCGGAGACTCCAGACTCCTGAAAGACTCCTAACAGACTCCGAAGACTCCCGACTCCAAGACTCCTTCAGACTCCAGACTCCTAGACTCCAGACTCCGTAATCAGGGAGACTCCAGACTCCTAAGACTCCATAGACTCCAGACTCCCACGAGACTCCAGACTCCAGACTCCGGAGACTCCAGACTCCTCTTGAGACTCCAACGCAATTAATTGTAGACTCCCCAAGACTCCCCAGACTCCTAGACTCCGCGGAAGACTCCAAGACTCCCTGTCGGGGGATAAGACTCCTAGACTCCGAGACTCCGACCGGAGACTCCTAGACTCCAGACTCCTGAAGACTCCACAGACTCCCTACAGAGACTCCTTTAGACTCCAAGGCCTAGACTCCAGACTCCTTCAGCAGAGTTAAGACTCCCTCATAGACTCCTAGACTCCCAAAAGACTCCAGACTCCAGACTCCCTGAAGAGACTCCAAGACTCCCTATTTTAGACTCCAGACTCCTGAGTAGACTCCATACAGACTCCAAGACTCCAGACTCCGCCTAGACTCCTCTAGTAGACTCCCGGAGACTCCAGACTCCAAGACTCCGACTCAGACTCCCCTTAGACTCCCAGACTCCATAAGGGAAGACTCCAGACTCCAGACTCCCAGACTCCGTCCAGCAAGACTCCAGACTCCAGACTCCAGACTCCCTATAGACTCCAGACTCCTAGACTCCCAAAGACTCCCGCATAAGTAGACTCCAGACTCCCAGACTCCTACCTTAGACTCCCGTTAGTAGACTCCTGTCGAGACTCCCTAGACTCCAGACTCCTAGACTCCAGTGGAGACTCCAGACTCCCGAGACTCCGATAAGACTCCAGACTCCCAGACTCCGGCAGACTCCCAGACTCCCGCGGTTCCCAGACTCCAGACTCCGCAGACTCCGTAGACTCCCAGACTCCTCAAGACTCCAGACTCCTTTAAAGACTCCTAGGAGACTCCAGTACTAGACTCCCAGACTCCGGAGACTCCAGACTCCAACTGAGACTCCTCAGACTCCGAGTAATGTATAGACTCCTAGACTCCTTAAATCAGACTCCAATAAGACTCCAGACTCCAACGGCGGAGATATAGACTCCTGGCTGATTCAGACTCCTAGTAAGACTCCACGTAGACTCCCAAGACTCCAGACTCCGAAAGACTCCAGACTCCAGCAGACTCCAGACTCCCCAGACTCCTAGACTCCAGACTCCGACAGACTCCACGTTAGACTCCGGAAAGACTCCTGAGACTCCGTGCATATAATTGAAGACTCCCCGTAGACTCCGTAAGACTCCAGACTCCCGAGAGACTCCAGACTCCGCACGTGAAGACTCCAGACTCCTAGACTCCAGAGACTCCAGACTCCAAGACTCCGAGGTAAGAGACTCCATGTAGTGAGACTCCACTTAGACTCCACACAAGACTCCACAGACTCCCAAGACTCCTAGACTCCTCTACGAGCCCAACTACCCAGACTCCAGCGCAGACTCCGTAGGCCGAGACTCCAGACTCCCAGACTCCTGAAGACTCCTACAGACTCCTTTAGACTCCGCTAACCTAGACTCCAGACTCCTCTCCAACCGCAGTAAGACTCCGAGACTCCTGAAGACTCCAAGACTCCTAGACTCCAGACTCCTGAGACTCCAGACTCCGAGACTCCTGTAGGAATAGTGAGCTAGACTCCAAACCGAGACTCCCGAACAGATTTGGAGACTCCAAGACTCCTGAGACTCCCGTAGAGACTCCAAGTAGGCGCTGAGACTCCAAGACTCCTCTTCAGACTCCAGACTCCGAACCGCCAAGACTCCAGACTCCAGACTCCGACTAAGACTCCTAATAGAGTGAGAGACTCCGCTAAAGACTCCAAGACTCCCACCTGGTCAGACTCCAGACTCCAGACTCCCCAGACTCCAGACTCCGTAGACTCCAGACTCCGTGTTCAAGACTCCGAAGACTCCAGAGACTCCGAGACTCCTAGACTCCCACGGTTCGAGACTCCGAGAAAGACTCCGCAGACTCCGTAGACTCCTAAGAGACTCCCTAGACTCCGAGACTCCCGAGACTCCAGACTCCAGACTCCAAGTTAGACTCCAGACTCCCTGAGACTCCATATCCTGATTGACATAGACTCCATAAAGACTCCAGACTCCTAGACTCCGAAGACTCCGAGACTCCATAGACTCCAGACTCCCGTAGACTCCAGACTCCTTAAGACTCCCGAAAATTATAGACTCCTAGACTCCTAGACTCCCGCCACACCAGACTCCGAGCAGACTCCAGACTCCCTTCAGACTCCAGACTCCTGTGAGACTCCAGAGAACTAGACTCCCGGCGTAGACTCCAAGACTCCGTCCCCGCCAGACTCCGAGACTCCGTATGAGACTCCAGACTCCGTAGACTCCGAGACTCCGGCGTAGACTCCCGGAGACTCCTCAAGACTCCAGACTCCTAGACTCCTAGGCCGGCTGGCAGACTCCGGTAGGTGGTAGACTCCAGACTCCAGTCAGACTCCTAGACTCCAGACTCCCGCCAGACTCCTTAGACTCCAGACTCCATATGAGACTCCTCCGAGACTCCAGACTCCGAGACTCCCGTTAGGTGAGACTCCGTACCCCACACAGGAGACTCCTATATTGGTTACAGACTCCAGACTCCAGACTCCAGGAGACTCCTGCTCAGACTCCTGGAAAGACTCCCAGACTCCCAAGACTCCCCTAAGACTCCAGACTCCTAGACTCCTGAGACTCCCGTAGACTCCATCCTGAGACTCCATCAAGACTCCAGACTCCGCAGACTCCAGTAAAGACTCCGAGACTCCCCTAGACTCCACGTTCAGACTCCCTGCATGCTGAAGACTCCTTTGTAGACTCCAGACTCCTAAGACTCCGATCCAGACTCCGTGGAAGACTCCAGACTCCCTTGAGACTCCCGATGGAGAGACTCCGGAGACTCCAGACTCCGCTAGACTCCGTAGACTCCGCGAAAGACTCCTAGACTCCAGACTCCCGTAGACTCCAAGACTCCGGAGACTCCAGACTCCGTTTCAGACTCCAGACTCCATAGACTCCAGACTCCTAGACTCCTCAGACTCCGTTCAGACTCCCCCAAGACTCCGAGACTCCAAGAGACTCCAGAGACTCCATTAGACTCCAGACTCCATGCAGACTCCGCCCCAGACTCCCAGACTCCCTAGACTCCCCAGAGACTCCAGACTCCAGACTCCAGACTCCAGACTCCCCAAGACTCCAGACTCCAGACTCCAGTCCAACAGACTCCCCAAGCAAGTAGACTCCCTAGACTCCAGACTCCGGTGAGACTCCAAGACTCCTAACTATCGAGACTCCTTAGACTCCAGACTCCACAGACTCCAGACTCCGACCGGAGACTCCGACAGACTCCAGACTCCGCGTAGACTCCTAGACTCCTATAGACTCCAGGTAGACTCCAGACTCCTAGACTCCCACAGAGACTCCGGAGAAAGACTCCCCCAGACTCCATAGACTCCAGTGACGGTACTGGTAGACTCCCAGACTCCTAAAGACTCCAAGACTCCGAAGACTCCCCACAGACTCCGTTAGACTCCAGACTCCCGACTGAGACTCCACAGACTCCAGACTCCGATGAAGACTCCGAGAGACTCCAGACTCCAGACTCCCGGTAGACTCCAGGAAAGACTCCTGAGACTCCCAGACTCCCCTACAGACTCCGTAGACTCCAGTAGACTCCTTAGACTCCAGACTCCCAGACTCCCATAGACTCCACAACGAGACTCCAGACTCCGAGACTCCAAGACTCCCGCAGACTCCAGACTCCACCAGCTTCACAGACTCCCTCAGACTCCGAGACTCCTAGACTCCAGAGACTCCGAGACTCCAGACTCCCAGACTCCTAGACTCCTGATAGACTCCAGACTCCAGTAGACTCCCAGACTCCACACCCTCACTCAGACTCCAAAGACTCCTAGACTCCAGACTCCTAGACTCCAAGACTCCGAGACTCCAGACTCCGCAGACTCCAAAGACTCCTCGAGACTCCAGACTCCAAAGACTCCAAAGACTCCCAGACTCCTCAGACTCCAGACTCCAGACTCCAGACTCCAGACTCCAGACTCCAGACTCCTAGGAAGACTCCCTTGAGACTCCTGAGACTCCAGACTCCAGACTCCGAGACTCCCTCAGACTCCGATAGACTCCGGAGACTCCAGACTCCAACTTTTGAAGACTCCAGACTCCAGACTCCCACGATCCGGCAAGACTCCCAAGACTCCGTGGAGACTCCAAGACTCCAGACTCCAGACTCCCAGTTGTCGGCCGAGACTCCTAAGTAGACTCCAGACTCCCAGACTCCAAGACTCCTAAGACTCCCTCAGACTCCTAAGACTCCCGGAGACTCCATAGACTCCATGGGAGTTGAACTTGAGATCTGACAACTCTAGACTCCTCGGGATAGACTCCGACAGACTCCGTAGACTCCCTCTCTCTTTAGACTCCAGACTCCGAGACTCCAGACTCCAGACTCCTACGTAGACTCCCGTAGACTCCGCAAGACTCCTGAAGACTCCAGACTCCAGACTCCAGACTCCTCGCGGAGACTCCGGAGACTCCCCAGACTCCAGACTCCATACAGACTCCAGACTCCGGGAGACTCCGGTAGACTCCATCCAGACTCCTAAGACTCCAGACTCCCAGACTCCTGAGACTCCTGAGACTCCAACAGACTCCAAGACTCC\n",
      "\n",
      "Positions = 41 48 86 93 111 126 182 198 215 239 264 285 318 379 386 401 408 512 527 595 633 677 684 691 709 716 723 742 759 786 820 848 855 870 877 884 937 986 1048 1079 1150 1174 1181 1216 1304 1319 1336 1370 1386 1393 1400 1407 1414 1429 1471 1518 1545 1654 1694 1716 1738 1753 1768 1811 1834 1874 1909 1916 1953 2005 2012 2055 2081 2102 2156 2172 2200 2353 2374 2381 2477 2484 2491 2506 2579 2595 2659 2722 2738 2762 2781 2814 2847 2906 2922 2929 2936 2981 2988 3031 3085 3109 3133 3140 3230 3256 3263 3299 3395 3428 3444 3459 3474 3481 3521 3561 3568 3575 3598 3613 3632 3639 3734 3797 3812 3836 3861 3879 3886 3902 4047 4110 4168 4175 4217 4255 4296 4357 4364 4394 4401 4408 4426 4467 4530 4545 4557 4584 4634 4677 4707 4737 4816 4893 4910 4917 4927 4951 5042 5060 5082 5097 5106 5226 5253 5313 5376 5392 5532 5555 5562 5636 5643 5659 5675 5705 5813 5820 5839 5890 5931 5948 6025 6043 6061 6133 6189 6241 6248 6267 6294 6324 6395 6402 6409 6471 6529 6545 6618 6658 6700 6746 6780 6799 6815 6879 6898 6956 6963 6970 6977 6994 7001 7008 7049 7100 7116 7146 7182 7193 7252 7330 7359 7388 7395 7413 7463 7482 7520 7553 7604 7621 7655 7662 7716 7747 7782 7824 7831 7838 7845 7852 7859 7898 7905 7949 7972 7979 8033 8040 8079 8237 8252 8259 8308 8315 8322 8360 8378 8425\n"
     ]
    }
   ],
   "source": [
    "fnm = 'Input_PatternMatching.txt'\n",
    "with open('Course1_Data/'+fnm,'r',encoding='utf-8') as file:\n",
    "    Pattern = file.readline().strip()\n",
    "    Genome = file.readline().strip()\n",
    "positions = PatternMatching(Pattern, Genome)\n",
    "\n",
    "print(f'Pattern = {Pattern}\\n')\n",
    "print(f'Genome = {Genome}\\n')\n",
    "print('Positions = '+' '.join([str(x) for x in positions]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee7f57c8",
   "metadata": {},
   "source": [
    "### Clump Finding Problem: Find patterns forming clumps in a string.\n",
    "\n",
    "  * Input: A string Genome, and integers k, L, and t.\n",
    "  * Output: All distinct k-mers forming (L, t)-clumps in Genome.\n",
    "  \n",
    "(L, t)-clump refers to the k-mer appearing at least t times in a window of length L in the genome.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a5e410f",
   "metadata": {},
   "source": [
    "To solve this problem, for every possible k-mer we find all the positions in the genome it appears.  \n",
    "Then for each k-mer, we test if there is a window of L where the k-mer appears at least t times.  \n",
    "To do this efficiently, we convert the k-mers into base-10 number, and then convert them back at the end. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "90279c27",
   "metadata": {},
   "outputs": [],
   "source": [
    "Alphabets = {'A':0,'C':1,'G':2,'T':3}\n",
    "\n",
    "def PatternToNumber(Pattern):\n",
    "    import numpy as np\n",
    "    Pattern = list(Pattern)\n",
    "    Pattern.reverse()\n",
    "    Base4 = [Alphabets[i] for i in Pattern]\n",
    "    FourToTen = [4**i for i in range(len(Pattern))]\n",
    "    Number = sum(np.array(Base4)*np.array(FourToTen))\n",
    "    return Number\n",
    "\n",
    "def NumberToPattern(Base10, k):\n",
    "    Alphabets = {0:'A',1:'C',2:'G',3:'T'}\n",
    "    Base4 = []\n",
    "    Pattern = []\n",
    "    for i in range(k):\n",
    "        Base4.insert(0,Base10%4)\n",
    "        Base10 = Base10//4\n",
    "        Pattern.insert(0,Alphabets[Base4[0]])\n",
    "    return ''.join(Pattern)\n",
    "\n",
    "def ClumpFind(Genome,k,L,t):\n",
    "    \n",
    "    def FindClump(positions,k,L,t):\n",
    "        for i in range(max(0,len(positions)-t+1)):\n",
    "            if positions[i+t-1]-positions[i] < L-k+1:\n",
    "                return True\n",
    "        return False\n",
    "    \n",
    "    PatternsPos = [[] for i in range(4**k)]\n",
    "    kMers = []\n",
    "    for i in range(len(Genome)-k+1):\n",
    "        PatternsPos[PatternToNumber(Genome[i:i+k])].append(i)\n",
    "    for i in range(len(PatternsPos)):\n",
    "        if FindClump(PatternsPos[i],k,L,t):\n",
    "            kMers.append(i)\n",
    "    kMers = [NumberToPattern(i,k) for i in kMers]\n",
    "    return kMers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "aa178ea9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genome = GCGGTTATGCACCGTTCAAATTAGCAAACCACTAAGCGACGTAGTCTGGATTGATTTCTCCCTACCAGTGACCCAAGACGCGTTAGTGAGTTAAGTTCATATCCAGTACCTGCCGCCCTCTGTACTTGGGCGTCCGATTCGCATGCTTACTCAGGTGGAGGACACGATAATCTGATTAAACTGAGCTAAACCAGGTGGAACCAGAAACCAGGTGGGGAGTCTCGCTTCAAGCCGTTCTTGCGATCAAACCAGGTGGTCCATTATGAAACCAGGTGGCTAAACCAGGTGGTCCAGATCCTCGAATGATGTCGGTGCACATCAAAACCAGGTGGGGTGGTGGAACGTAAAACCAGGTGGCATAAACCAGGTGGGCCGGTTCGTAAACCAGGTGAAACCAGGTGGGGTGGAAACCAGGTGGGTTACAAATTACGTTGAGATGGCCCAAACCAGGTGGTGGGCTTCACCCATGTCAACAAACCACCCTATGGAACTAAACCAGGTGGAACCAGGTGGTGAAGGCTTATCCTCAGGAAAAACCAGGTGGAGGTGGTGAAATAAAACCAGGTGGACCAGGTGGATAACCCTCGCCTCGCTTCTCAACCGAGACCTGGATAAACCAGGTGGGGTGGTCCACCGATTTTTGAGACACTAGAAACCAGGTGGGCGGGGAAACCAGGTGGCAAACCAGGTGGGGTGGACGGAAACCAGGTGGATATGTCATAAAACCAAACCAGGTGGTGCACCCCCATGGTGTGTCTTATCCGTGCGTATAAACCAGGTGGTCGCACGGCTTCCACTTGCTGAGAATAGGCCCGCAGGGTCAGTGCCATGCCCTCCGTCACTCGATATGTGTTGTAAGAGTGGTTACCCCTTCATTGAAGTCGCCCACAGCCCCACCTGCATTGCTAGACTATCACCCTACAGTAGGCCTTTTCGCCTTCTTCAAGCAGCAATCTCTTATCCGCGGATGGGCGCGGCGAGCGTGGCGTCCCCGAACATTTTTACCTAACGTGTTTTGTTGGCCGCAAGCCTTCCCTCTAGTCCACCTCAGCCATTCAGCCTAGTAGCTTTCAAGCCGAGCCTTCCATATCTAATGGACCGTCCAGAATTTCACACGTTTCACAGGGCTGTGTTCGACCGCCCGTAATGCTGTTTCACAGGCGATCGCCTTGCGGTTTTTTCACAGATCGCAGCCGATGGACATGCCAACTCGATTTTCACAGAGTTTTTCACAGCGGTTTCACAGCACAGCAGTGATTGTTTCACAGCAATTTTCACTTTCACAGGGGCCCTTTTCACAGCTCAGGGCTCTTTTCACTTTCACAGTTTCACAGCGCTCCTTTCACAGAGCGGGGAAATTTAAGGGAACACTCAAGGGAACAAGGGAACACACAAAGGGAACACAACACAACACATAAGGGAACACTTTCACAGAACACAAAAGTCCGAAATCATCAGCGGCGAAGGGATTTCACAGACAGACACTTTCACAGCGCATTTCACAGATACGTACTTTCACAGGCGTACTTTCACAGACTTTCACAGAGGACAAGCTCAATTTTCACAGACAGGCTGGATAAATTTCACAGCGGTAAGGGTTTCACAGCACACATAAGGGAACACGAATTTCACAGCAGGGAACACCTCTACGAGTAATCTATTACTCTACCTACTGAAGGGAACACACCGAAGACCTACTATTACCTATTACTCTTAAAGGGAACACATTACAAGGGAACACACTCTCTCGTCATATCTCACCTCTCTATTACTCTTAAGGGAACACCTTCTCGATCAACCTATTACTCTATGGAGATAGAGATATTCCAGACATATGGAGATAACATGGAGATATGGAGATAATGGAGATGGAGATAGCTCTTATATTTATCCTATGGAGATATGATACTATTAATGGAGATAATTCTAATGGAGATATAATTACTCTAAGAGGATGGGATCTCGGGCTATTACTCTAATGGAGATAAGCACTATTACTCTAGGAAATGGAGATATGTCAATGGAGATATGTAATGGAGATAGAGGGAGATGGAGTCGCCATTTCATAATCGCCATTTCATAGTTCAGGAATCGCCATTTCCGCCATTTCTAAGATGGAGTCGCCATTTCTACGTATGGAGATAGGATCGCCATTTCATACGACCCGTTGGATATCGCCATTTCCTCGCCATTTCTGGTGACATTTCTCGCCATTTCATTTCTGGAGATAGATGGATCTCGCCATTTCATAGGAATCGCCATTTCCACGTAGGGGGGGCCACAATCCGTAGGTCGGAATTCAGACTCGCCATTTCCCATCGCCATTTCTTCACCTGTATGCCGATCCCTTCGCCATTTCTCATGGAGATAACTCTCTCTCGCCATTTCTCGCCATTTCCATTTCACTCTCATTCGCCATCGCCATTTCCATTCGCCATTTCATCGCCATTTCTTCAGGATAAGATATCGCCATTTCGACTCTCATTCGCATACTGACTCTCATTCTCATCTCGCCATTTCTCATCTGACTCTCATCCTGGGGGAAACTTGCGACTCTCATCACACTTCCGTCGACTCTCATACTGGCGGATAGCATAGGAGCCATTTAAAGACTCTCATTCTCATTCGAGACTCTCATTCAAATCCTACGAGGACTCTCATATAGACTCTCATATCATTACGAGGACTCTCATATACGAGCCATGCATGTGGCGACGACTCTCATCTACGAGCCATGCAAGCAGAATCTACGAGCGACTCTCATTACGAGCCATGTGACCGTACGAGCCATGCATGCATGCCATGCTGACTCTCATCGAGTACGAGCCATGGAAGTTCTTGTTGGTTCGTAGCCCAAGAGCTGAAGTTACGAGCCTACGAGCCATGAAGTTACTTTTACGAGCCATGAAGCTTACGATACGAGCCATGCGAGCCATGCATCCGCGCTACGAGCCATGTTCCAGTACGAGCCATGTTAGTTGCTGAAGTTAAGTTTGGCGCTGAAGTTTGTACGAGCCATGTGCCCGCTGAAGTTTGTTGTACGAGCCATGCATGCTGAAGTTAATGGCTGAAGTTAGCGTTTGCGGGCAGATCCTCATTCTACGATACGAGCCATGCCATGCAGCTGAAGTTAAGTTGGGTTACGAGCCATGCGAGCCATGTGAAGTACGAGCCATGCTGGCTGAAGTTGTTTGTGCTGCTGAAGTTGCTCTTGTCTCTAGCTGAAGTTGCCAACAGGGCTGAAGCTGAAGTTTAAGCTGAAGTTGCGAGCAGGCTGAAGTTATCGGATTGGGGCTGAAGTTCAACCTCCCGTCCCCCCACACTATATTCCCGTCCCCCCCCGCGCACGCGCCGTCTCCCGTCCCCCCTATCCCGTGCGCACGCGACGCGATCCCGTCCCCCCAGAGTGCGCGCACGCGTCCCCCTTCCCGTCCCCCTCTCCCGGGCGCACGCGTCGCTCAACATTTCCGCGCACGCGTCGCGCACGCGGGCGCACGCGGGTCCCGTCCCCCCCCCTCTTCGGCGCACGCGGAATTCCCGTCGCGCACGCGTCCCGTCCCGCGCACGCGTCGCGCACGCGACTGCCCTAACCAACAGTGCGCACGCGCCGGTAACCCGGTAACCCGGTAACCGCGCACGCGGGCGCACGCGCGTAACCCGCGCACGCGCCGCGCACGCGGCCCGGTTCCCGTCCCCCCCGGTAACCCGGTAACTCCCGTCCCCCGTAACCCGGTGCGCACGCGCCCGGCGCACGCGGAGCGCACGCGCCCCCCCCGGTAATAGCGCACGCGCCCGGGCGCACGCGCCCGGTAACCCGGTAACCCGGGCGCGCGCACGCGGCGGCGCACGCGGCGCACGCGGCGCACGCG\n",
      "\n",
      "k = 11, L = 566, t = 18\n",
      "Clumped Kmers = AAACCAGGTGG\n"
     ]
    }
   ],
   "source": [
    "fnm = 'Input_ClumpFinding.txt'\n",
    "with open('Course1_Data/'+fnm,'r',encoding='utf-8') as f:\n",
    "    Genome = f.readline().strip()\n",
    "    k, L, t = [int(x) for x in f.readline().strip().split()]\n",
    "\n",
    "kMers = ClumpFind(Genome,k,L,t)\n",
    "print(f'Genome = {Genome}\\n')\n",
    "print(f'k = {k}, L = {L}, t = {t}')\n",
    "print('Clumped Kmers = ' + ' '.join(kMers))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3dcefeb0",
   "metadata": {},
   "source": [
    "### Frequent Words with Mismatches and Reverse Complements Problem: \n",
    "Find the most frequent k-mers (with mismatches and reverse complements) in a string.\n",
    "\n",
    "  * Input: A DNA string Text as well as integers k and d.\n",
    "  * Output: All k-mers Pattern maximizing the sum Count<sub>d</sub>(Text, Pattern)+ Count<sub>d</sub>(Text, Pattern<sub>rc</sub>) over all possible k-mers.\n",
    "\n",
    "I solve this problem by first obtaining all patterns that are within the Hamming distance of d from any k-mer substring in the text. Then, for each of them, I count Count<sub>d</sub>(Text, Pattern)+ Count<sub>d</sub>(Text, Pattern<sub>rc</sub>) and find the maximum. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "42590637",
   "metadata": {},
   "outputs": [],
   "source": [
    "def FrequentWithMismatch(Text, k, d):\n",
    "    \n",
    "    # Create a list of patterns that are length d \n",
    "    def SimilarPatterns(pattern, d):\n",
    "        import itertools\n",
    "        similars = []\n",
    "        C = list(itertools.combinations(range(len(pattern)),d))\n",
    "        for comb in C:\n",
    "            similars = similars + ModifyPattern(list(pattern), list(comb), [])\n",
    "        return list(set(similars))\n",
    "\n",
    "    # Use recursion to modify patterns at given indices \n",
    "    def ModifyPattern(pattern, indices, patternslist):\n",
    "        if len(indices)==0:\n",
    "            patternslist.append(pattern.copy())\n",
    "            return\n",
    "        else:\n",
    "            alphabets = ['A','C','G','T']\n",
    "            for s in alphabets:\n",
    "                pattern[indices[0]] = s   \n",
    "                ind = indices.copy()\n",
    "                del ind[0]\n",
    "                ModifyPattern(pattern,ind,patternslist)\n",
    "            return [''.join(s) for s in patternslist]\n",
    "    \n",
    "    # Obtain the reverse complmeent of the pattern\n",
    "    def ReversePattern(pattern):\n",
    "        Reverse_dict = {'A':'T','T':'A','G':'C','C':'G'}\n",
    "        revpattern = [Reverse_dict[i] for i in list(pattern)]\n",
    "        revpattern.reverse()\n",
    "        return ''.join(revpattern)\n",
    "    \n",
    "    # Count the number of approximate matches of the pattern in the Text\n",
    "    def ApproxMatching(Pattern, Text, d):\n",
    "        \n",
    "        def HammingDistance(seq1, seq2):\n",
    "            x = 0\n",
    "            for i in range(len(seq1)):\n",
    "                if seq1[i] != seq2[i]:\n",
    "                    x = x+1\n",
    "            return x\n",
    "        \n",
    "        matches = []\n",
    "        k = len(Pattern)\n",
    "        for i in range(len(Text)-len(Pattern)+1):\n",
    "            if HammingDistance(Text[i:i+k], Pattern) < (d + 1):\n",
    "                matches.append(i)\n",
    "        return matches\n",
    "\n",
    "    AllPatterns = []\n",
    "    matches = []\n",
    "    # Get all patterns that are d away from any k-mer substring in the text\n",
    "    for i in range(len(Text)-k+1):\n",
    "        AllPatterns = AllPatterns + SimilarPatterns(Text[i:i+k], d)\n",
    "    AllPatterns = list(set(AllPatterns))\n",
    "    AllPatterns = [''.join(pattern) for pattern in AllPatterns]\n",
    "    for pattern in AllPatterns:\n",
    "        matches.append(len(ApproxMatching(pattern, Text, d))+len(ApproxMatching(ReversePattern(pattern),Text,d)))\n",
    "    MaxInd = []\n",
    "    MaxVal = max(matches)\n",
    "    for i in range(len(matches)):\n",
    "        if matches[i] == MaxVal:\n",
    "            MaxInd.append(i)\n",
    "    return [AllPatterns[i] for i in MaxInd]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "372a6cfb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genome = TAACATCATTTAATAATTTTTTAATAAAGTTTTAAAGAGCATTTTTTAATAAAGTAATAATAATTAACATTAACATAAAACATAGCATCATTTAGTAAAAAATAATAAAGAAAGCATAGTTTTTAAAGAGAAAAAGTAATAATTAGCATAGCATCATAGTTCATAGTAATTTAATAATAATAACATAACATTAACATCATTAAAATTTTAGTAATAAAACAT\n",
      "\n",
      "k = 5, d = 2\n",
      "kMers with most frequent approximate matching = ATTAT ATAAT\n"
     ]
    }
   ],
   "source": [
    "fnm = 'Input_FrequentWordsMismatch.txt'\n",
    "with open('Course1_Data/'+fnm,'r',encoding='utf-8') as f:\n",
    "    Genome = f.readline().strip()\n",
    "    k, d = [int(x) for x in f.readline().strip().split()]\n",
    "\n",
    "kMers = FrequentWithMismatch(Genome,k,d)\n",
    "\n",
    "print(f'Genome = {Genome}\\n')\n",
    "print(f'k = {k}, d = {d}')\n",
    "print('kMers with most frequent approximate matching = ' + ' '.join(kMers))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d42277c",
   "metadata": {},
   "source": [
    "### Motif Search Problem\n",
    "\n",
    "Given a set of t DNA strings, we want to find the most likely motif shared by all strings, allowing for mismatches. The optimization function is the sum of hamming distance of all of the actual motifs from the consensus motif. The consensus motif is determined by the profile matrix constructed from the motifs in all of the strings. The challenge here is that the choice of the motifs itself changes this probability. Thus, we need to use an EM algorithm to iteratively find the solution.\n",
    "\n",
    "We approach this problem in two different ways: Greedy search and Randomized search (Monte Carlo). For the randomized search, we also test the Gibbs sampling method."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cdb3f71",
   "metadata": {},
   "source": [
    "#### Shared utilities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "6793879e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def Score(motifs):\n",
    "    consensus = Consensus(motifs)\n",
    "    s = 0\n",
    "    for pattern in motifs:\n",
    "        s += Hamming(pattern,consensus)\n",
    "    return s\n",
    "\n",
    "def MostProbable(k, Pattern, Profile):\n",
    "    alphabets = ['A','C','G','T']\n",
    "    Profile = {k:v for k,v in zip(alphabets,Profile)}\n",
    "    probs = []\n",
    "    for i in range(len(Pattern)-k+1):\n",
    "        p = 1\n",
    "        for j in range(k):\n",
    "            alphabet = Pattern[i+j]\n",
    "            p = p*Profile[alphabet][j]\n",
    "        probs.append(p)\n",
    "    idx = probs.index(max(probs))\n",
    "    return Pattern[idx:idx+k]\n",
    "\n",
    "def Profile(motifs):\n",
    "    N = len(motifs)\n",
    "    k = len(motifs[0])\n",
    "    Usage = [[1/(N+4)]*k for j in range(4)]\n",
    "    alphabets = {'A':0,'C':1,'G':2,'T':3}\n",
    "    for pattern in motifs:\n",
    "        for i in range(k):\n",
    "            row = alphabets[pattern[i]]\n",
    "            Usage[row][i] += (1/(N+4))\n",
    "    return Usage # 2D list\n",
    "\n",
    "def Consensus(motifs):\n",
    "    import numpy as np\n",
    "    k = len(motifs[0])\n",
    "    alphabets = {'A':0,'C':1,'G':2,'T':3}\n",
    "    alphabets = {v:k for k,v in alphabets.items()}\n",
    "    profile = np.array(Profile(motifs))\n",
    "    consensus = []\n",
    "    for i in range(k):\n",
    "        consensus.append(alphabets[np.argmax(profile[:,i])])\n",
    "    return ''.join(consensus)\n",
    "\n",
    "def Hamming(kmer1, kmer2):\n",
    "    d = sum([i!=j for i,j in zip(list(kmer1),list(kmer2))])\n",
    "    return d"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3eade5f1",
   "metadata": {},
   "source": [
    "#### Greedy Search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "2b2f0129",
   "metadata": {},
   "outputs": [],
   "source": [
    "def GreedyMotifSearch(DNA, k, t):\n",
    "    \n",
    "    BestMotifs = [pattern[0:k] for pattern in DNA]\n",
    "    N = len(DNA[0])\n",
    "    for i in range(N-k+1):\n",
    "        motifs = []\n",
    "        motifs.append(DNA[0][i:i+k])\n",
    "        for j in range(1,t):\n",
    "            profile = Profile(motifs)\n",
    "            motifj = MostProbable(k,DNA[j],profile)\n",
    "            motifs.append(motifj)\n",
    "        if Score(BestMotifs) > Score(motifs):\n",
    "            BestMotifs = motifs\n",
    "    return BestMotifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "dde75d1a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "k = 12, t = 25\n",
      "Genomes = \n",
      "CTATGGGGTATTTTTGTGCGGGTATTAAAGTCGGATAACACCCACCATTACTTGCGTAACCATGGCTGCTAGAGATCTTGTAGCCACGTTTGACCTGTGTGAACTGTACCACGCTCGCCCCAATCGGCTTCCGCCACGCTGTTAAGATCCTACCGG\n",
      "CTGCAGCATAAGGACAACGACTTATCATTTCACCGCCGCCGCGCTACACAATAGGCTCACAGAGCTGACTGCCCTGTCGATTGTCCACAGGCAGGCCCCAATGGGCAAATAGCGAATAGTACCAACCTACGATGTGTTCCGCGGAAGTTACATACA\n",
      "TTGTTTGCCTCCGCTTCTTCGCCCAGTTAGTATTGCTTATAATCCCCGGATGCCATTTCGCAGTGACGATGACGATGGGCTACCCTTGGCCGATTCCTTTCGATGAGCCCTGATGAATCTGCGCGGAATCGGAGGGGGTTAATGGCACGCTAAACT\n",
      "GGTTCATTCGTTGTGTTGAAAAAGGGCACGTCGCCTCAATGGGCTCACGCATAGCTGTAGAGATTCGCGGTACCTGCCTTATAGAGCACCCACCTAAGGTGAGGCCCTATTTCGTAGATTACCTCTACATCAGTACTTATTAACGGGGGAAATACT\n",
      "TTAAGGCCGATAAGTATTTCCCCCGTTAGGTATTCTTACTCTCCTACAATTATCCCTGATGTTACGGTGATTCGTGCAACAACGTGGCAACAGTGTCGAGATGGGAGGCCATGGGCTGCCAGCGCTCATTTGAGTGTTTTCTTGTTCTGTGTCAAT\n",
      "CCGTTCACGTTCATAACAAGCAGGAAAACCTTTGGAGGCTTGGAGTCGGGTTGTTGAATGGTAGCTTAGTATCAATTGGCTGTCTTCTCACTTGCCGGTTTAGTAGACTCAGCTCCGTCTTAGCTCTTTTACGCATCCGAGCCTACCTCTGCGTCT\n",
      "TGCAGTATCGCATGCGCGCACTTCACGATGAAGCTAACGAGCGGACAACATGGTAGGACAGAGAACACTCTCCGTTCAAAATTCCGATTGGCTCTCTTAGGATTGTGCCAGACGGCATGAAAGGGGGATTATTCGATAGAGTGTCCCAACCGACTA\n",
      "ATAGGGACGTATTAGTCATGCTGTCGGACATGGCTTGAACACATTACATAGGGACCGTTGCGTTCAAGAGTTTTACCGTCACCTATGGAATATTTGTGGTGGCGTAAGTCCACCGGGCCAACTGTAAGTTTCACCTGCTAGACGCTATTGGCTCGC\n",
      "GCGGGCTAGGGACTTATTTAAACCACGTTCGTCGGCCGTTCCGTAACTTAACGCCGGAACTGCGGTACTCAACCAATGATACAAGTCCCCTACGCAGCCAAATCACTCGGGATGCGCACAGTCACACCTTGACTATGGGCTGTCCCGACTGCAGTG\n",
      "CGATGGGCTGACTTGCCAGTCTAGGCCAGAATCCCTAGTATTACTGTCGTTATTGTCCTCGATACGCACCTAACATTACTACGTATGGGTGTCTTGGCTTAGGGCGACCTATGCTCAGAATTAGAGATGCTCAACTGCAACTTCTGCAACGAAGTG\n",
      "GAGTCATGACGGGCTCTTGAGGCCCCGATGCGGCTGGGTAACAAGACGCGCAATGTTAGCACTACTTACCATAGTCTAGACGTGAACATGAATAGGGTGACAGCTGATGTAGACAACAGATGCCACGGATGCACTCCTCGTTAGCGATGGGCTAAC\n",
      "ATGAATTCTTTTAAAGTTTGTTCACGCTACTCTGCGTCTCCTAGGTCTCCATAGGCTGCCTCCCCTTAGGGAGGTAGAAGAACATAAATTACCTCGAATCCGGCATCTCGTTTGAGTCGTACCGTCATCAGAGGAAAGTGTCGGGATAAAACGGGA\n",
      "AGCGGAGGGTGATTGGACGCCTTAGAGATGCAGAGTAAGTGTGATTATCTAAATTAACAGTATTTGACAGTGTCACACGCTACAGTTTCCCGAATCTAAAAGACATATCAATCGGCTTTCACTCAGTCACCTTTAAGTACCAACGCCCGGTCGTAA\n",
      "GTGACTAATATCAGAGGTATAGAGATTCATTGGTTGAAACGAGCCGTTCTATTGGCTGCCGGAGAATTAATCGTGGACCATACATTGGTGCCACCCGCGGTCGGCTTGTGGTTTCATTGGAGGGCGACTGTGCAATACCACAGCACACCAGTGTCT\n",
      "TGACCCGTTACACATTTGCAGCTGGTTTTTTCCACATGTGTCATGCTCCAGCGGTTGATCACTGCAGGGAACACTAGCGAAAAAGTGCTTACGAGTCAACTTTCATATATCACACAGACGGCCAGGACATTTCCATCGGCTCACCCGGTGGGATTA\n",
      "CGCTCTCTGGTTGTGTAATGGAACGAAGCCCCCATGGGTTGCAGGAGTCGGCGACACCAACAAACAGCATACCAACGGCCAAGTATGCTCTACATCCGATGGGCTCGCCCTGGACCCGATGTAATTACTCGATCGGAACCCTTATACCCAGCGTGT\n",
      "GACACCTGGCCTGCCGAGGAAGAGAGAAACACGATCTCACCCCACTTCCGTTAATAGAAAGGTGTATGCTTGGACCGGTAACCTCGGCTTAGTTTGGGCTGTCGTCATCGATGGGCTGGCCGGAATCGAGGGAGTAGCCATGTCGCTGTCTCGATC\n",
      "GGACAGTCATTGAGCTATTGGGTACGCCTTCGAGTTTGATGGTCTCTCCTCGAGCCCTGTCAACTGAAGTAGTAGTTATGGCCAGCCCCAGCTGCCCTGCCATACACATCGCCGCGATTGATGTCCATTTTTCCATTGGCTTGCTCTGAAAAGTAA\n",
      "ACGTGGCAGAAGGCCTCGATCAAGTTCAAGAGTCTAATTGGTTCGTGGGTTGTTACAACAATTATTTTCTGACCATGGGCTCACAGACGCGCACATGGACGATAACGGCGGGCATAGGTCTTCTTGTCGAAGGAAGCCTCCCATGGGAAGCGACCG\n",
      "CTATAAATACGCCTCGAACATGCGGCACCAGTTTTCTCGGATTCTGGCCTGGAGGTGAGTTGTACCATCAAGATGTAGTGCTACCTATCGGCTGTCTGGTCTGAAATTACCCTTTAGATCCCAATAGTGCGGGAATACCGCTTTGAATTTACTCAT\n",
      "CGCGATCATCTCGAGCTACTTATATATACTCTTGTAGCCGATTGTCCTCTATGGGCTAGCTTCTATACTGAGTCAGTATCCGCCCCGTCAGGATCTATCCAGGTAGGCGACCTTACAACCACTGGCAAGGTAACCTGTTGTGGGCTCGTAGCGCAA\n",
      "TCATGGGTCGTTTGGCACCTGCAGACAATAGTACAGCTCGTACCACTTGCAGAGGTAAATTCCATCGGCAGTCCATCGGCTCCCTGATGTCTGAATAGCTAACCAAAATCTCCCACAAGATTTATTGCCGTGACCTATGGCCTTTCACGCTGGGCT\n",
      "GATCTGATTCGTAGTAGCTAGTCCCGAAGCTCTTATGAACACGCTGATACGATACAATTGTTGATAAGAGGCCCAACCGCGATAGGATGCGGCTGATATTTGTTATGTCCATGGGCTTCCCCCTTAGGGATCCTAACCTCTGTTCGTATTATTCCG\n",
      "GAAGGTGCCATACCTTTTCCTGGTTCCGACCTTTCGTCTCGATTTAAAGCGTATCTCGCCAGCCGATGGTTCTTGGTCTTAGATAGGCGTGATGCATCTTTCAGACGGGGCCAGCGGCTGGGGAAATTAAGCCAATTGGCTGTCCACGGCGCGATG\n",
      "TCACGTGTTCTTATTGTGTTCCGGCGAAGTTCCGTTTGTCCCTGCGATATGAAGCAGTTATCTCCTAGAATTTCCCGATAAGTCGAAGCTGGTGTTCTATCTTTGCGGCCATTGGCTCTCCAGCATCCGTAAGAGAGACGCGCGAGAAAATCCCAC\n",
      "\n",
      "Best motifs = CAATCGGCTTCC CAATAGGCTCAC CGATGGGCTACC CAATGGGCTCAC CCATGGGCTGCC CAATTGGCTGTC CGATTGGCTCTC CTATTGGCTCGC CTATGGGCTGTC CGATGGGCTGAC CGATGGGCTAAC CCATAGGCTGCC CAATCGGCTTTC CTATTGGCTGCC CCATCGGCTCAC CGATGGGCTCGC CGATGGGCTGGC CCATTGGCTTGC CCATGGGCTCAC CTATCGGCTGTC CTATGGGCTAGC CCATCGGCTCCC CCATGGGCTTCC CAATTGGCTGTC CCATTGGCTCTC\n"
     ]
    }
   ],
   "source": [
    "filename = 'Input_MotifSearch.txt'\n",
    "Genome = []\n",
    "with open('Course1_Data/'+filename,'r',encoding='utf-8') as file:\n",
    "   k,t = tuple([int(i) for i in file.readline().strip().split(' ')])\n",
    "   for i in range(t):\n",
    "       Genome.append(file.readline().strip())\n",
    "BestMotifs = GreedyMotifSearch(Genome,k,t)\n",
    "print(f'k = {k}, t = {t}')\n",
    "print('Genomes = \\n' + '\\n'.join(Genome))\n",
    "print('\\nBest motifs = ' + ' '.join(BestMotifs))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e9b4ea9",
   "metadata": {},
   "source": [
    "#### Monte Carlon Randomized Search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "f8451265",
   "metadata": {},
   "outputs": [],
   "source": [
    "def MCMotifSearch(DNA, k, t, N_iter=1000):\n",
    "    \n",
    "    def RandomizedMotifSearch(DNA, k, t):\n",
    "        from random import randrange\n",
    "        N = len(DNA[0])\n",
    "        BestMotifs = []\n",
    "        for pattern in DNA:\n",
    "            i = randrange(N-k+1)\n",
    "            BestMotifs.append(pattern[i:i+k])\n",
    "        while True:\n",
    "            motifs = []\n",
    "            profile = Profile(BestMotifs)\n",
    "            for pattern in DNA:\n",
    "                motifs.append(MostProbable(k,pattern,profile))\n",
    "            if Score(BestMotifs) > Score(motifs):\n",
    "                BestMotifs = motifs\n",
    "            else:\n",
    "                return Score(BestMotifs), BestMotifs\n",
    "    \n",
    "    BestScore = k*t+1\n",
    "    BestMotifs = None\n",
    "    for i in range(N_iter):\n",
    "        score, motifs = RandomizedMotifSearch(DNA, k, t)\n",
    "        if BestScore > score:\n",
    "            BestScore = score\n",
    "            BestMotifs = motifs\n",
    "    return BestMotifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "12a3c272",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "k = 12, t = 25\n",
      "Genomes = \n",
      "CTATGGGGTATTTTTGTGCGGGTATTAAAGTCGGATAACACCCACCATTACTTGCGTAACCATGGCTGCTAGAGATCTTGTAGCCACGTTTGACCTGTGTGAACTGTACCACGCTCGCCCCAATCGGCTTCCGCCACGCTGTTAAGATCCTACCGG\n",
      "CTGCAGCATAAGGACAACGACTTATCATTTCACCGCCGCCGCGCTACACAATAGGCTCACAGAGCTGACTGCCCTGTCGATTGTCCACAGGCAGGCCCCAATGGGCAAATAGCGAATAGTACCAACCTACGATGTGTTCCGCGGAAGTTACATACA\n",
      "TTGTTTGCCTCCGCTTCTTCGCCCAGTTAGTATTGCTTATAATCCCCGGATGCCATTTCGCAGTGACGATGACGATGGGCTACCCTTGGCCGATTCCTTTCGATGAGCCCTGATGAATCTGCGCGGAATCGGAGGGGGTTAATGGCACGCTAAACT\n",
      "GGTTCATTCGTTGTGTTGAAAAAGGGCACGTCGCCTCAATGGGCTCACGCATAGCTGTAGAGATTCGCGGTACCTGCCTTATAGAGCACCCACCTAAGGTGAGGCCCTATTTCGTAGATTACCTCTACATCAGTACTTATTAACGGGGGAAATACT\n",
      "TTAAGGCCGATAAGTATTTCCCCCGTTAGGTATTCTTACTCTCCTACAATTATCCCTGATGTTACGGTGATTCGTGCAACAACGTGGCAACAGTGTCGAGATGGGAGGCCATGGGCTGCCAGCGCTCATTTGAGTGTTTTCTTGTTCTGTGTCAAT\n",
      "CCGTTCACGTTCATAACAAGCAGGAAAACCTTTGGAGGCTTGGAGTCGGGTTGTTGAATGGTAGCTTAGTATCAATTGGCTGTCTTCTCACTTGCCGGTTTAGTAGACTCAGCTCCGTCTTAGCTCTTTTACGCATCCGAGCCTACCTCTGCGTCT\n",
      "TGCAGTATCGCATGCGCGCACTTCACGATGAAGCTAACGAGCGGACAACATGGTAGGACAGAGAACACTCTCCGTTCAAAATTCCGATTGGCTCTCTTAGGATTGTGCCAGACGGCATGAAAGGGGGATTATTCGATAGAGTGTCCCAACCGACTA\n",
      "ATAGGGACGTATTAGTCATGCTGTCGGACATGGCTTGAACACATTACATAGGGACCGTTGCGTTCAAGAGTTTTACCGTCACCTATGGAATATTTGTGGTGGCGTAAGTCCACCGGGCCAACTGTAAGTTTCACCTGCTAGACGCTATTGGCTCGC\n",
      "GCGGGCTAGGGACTTATTTAAACCACGTTCGTCGGCCGTTCCGTAACTTAACGCCGGAACTGCGGTACTCAACCAATGATACAAGTCCCCTACGCAGCCAAATCACTCGGGATGCGCACAGTCACACCTTGACTATGGGCTGTCCCGACTGCAGTG\n",
      "CGATGGGCTGACTTGCCAGTCTAGGCCAGAATCCCTAGTATTACTGTCGTTATTGTCCTCGATACGCACCTAACATTACTACGTATGGGTGTCTTGGCTTAGGGCGACCTATGCTCAGAATTAGAGATGCTCAACTGCAACTTCTGCAACGAAGTG\n",
      "GAGTCATGACGGGCTCTTGAGGCCCCGATGCGGCTGGGTAACAAGACGCGCAATGTTAGCACTACTTACCATAGTCTAGACGTGAACATGAATAGGGTGACAGCTGATGTAGACAACAGATGCCACGGATGCACTCCTCGTTAGCGATGGGCTAAC\n",
      "ATGAATTCTTTTAAAGTTTGTTCACGCTACTCTGCGTCTCCTAGGTCTCCATAGGCTGCCTCCCCTTAGGGAGGTAGAAGAACATAAATTACCTCGAATCCGGCATCTCGTTTGAGTCGTACCGTCATCAGAGGAAAGTGTCGGGATAAAACGGGA\n",
      "AGCGGAGGGTGATTGGACGCCTTAGAGATGCAGAGTAAGTGTGATTATCTAAATTAACAGTATTTGACAGTGTCACACGCTACAGTTTCCCGAATCTAAAAGACATATCAATCGGCTTTCACTCAGTCACCTTTAAGTACCAACGCCCGGTCGTAA\n",
      "GTGACTAATATCAGAGGTATAGAGATTCATTGGTTGAAACGAGCCGTTCTATTGGCTGCCGGAGAATTAATCGTGGACCATACATTGGTGCCACCCGCGGTCGGCTTGTGGTTTCATTGGAGGGCGACTGTGCAATACCACAGCACACCAGTGTCT\n",
      "TGACCCGTTACACATTTGCAGCTGGTTTTTTCCACATGTGTCATGCTCCAGCGGTTGATCACTGCAGGGAACACTAGCGAAAAAGTGCTTACGAGTCAACTTTCATATATCACACAGACGGCCAGGACATTTCCATCGGCTCACCCGGTGGGATTA\n",
      "CGCTCTCTGGTTGTGTAATGGAACGAAGCCCCCATGGGTTGCAGGAGTCGGCGACACCAACAAACAGCATACCAACGGCCAAGTATGCTCTACATCCGATGGGCTCGCCCTGGACCCGATGTAATTACTCGATCGGAACCCTTATACCCAGCGTGT\n",
      "GACACCTGGCCTGCCGAGGAAGAGAGAAACACGATCTCACCCCACTTCCGTTAATAGAAAGGTGTATGCTTGGACCGGTAACCTCGGCTTAGTTTGGGCTGTCGTCATCGATGGGCTGGCCGGAATCGAGGGAGTAGCCATGTCGCTGTCTCGATC\n",
      "GGACAGTCATTGAGCTATTGGGTACGCCTTCGAGTTTGATGGTCTCTCCTCGAGCCCTGTCAACTGAAGTAGTAGTTATGGCCAGCCCCAGCTGCCCTGCCATACACATCGCCGCGATTGATGTCCATTTTTCCATTGGCTTGCTCTGAAAAGTAA\n",
      "ACGTGGCAGAAGGCCTCGATCAAGTTCAAGAGTCTAATTGGTTCGTGGGTTGTTACAACAATTATTTTCTGACCATGGGCTCACAGACGCGCACATGGACGATAACGGCGGGCATAGGTCTTCTTGTCGAAGGAAGCCTCCCATGGGAAGCGACCG\n",
      "CTATAAATACGCCTCGAACATGCGGCACCAGTTTTCTCGGATTCTGGCCTGGAGGTGAGTTGTACCATCAAGATGTAGTGCTACCTATCGGCTGTCTGGTCTGAAATTACCCTTTAGATCCCAATAGTGCGGGAATACCGCTTTGAATTTACTCAT\n",
      "CGCGATCATCTCGAGCTACTTATATATACTCTTGTAGCCGATTGTCCTCTATGGGCTAGCTTCTATACTGAGTCAGTATCCGCCCCGTCAGGATCTATCCAGGTAGGCGACCTTACAACCACTGGCAAGGTAACCTGTTGTGGGCTCGTAGCGCAA\n",
      "TCATGGGTCGTTTGGCACCTGCAGACAATAGTACAGCTCGTACCACTTGCAGAGGTAAATTCCATCGGCAGTCCATCGGCTCCCTGATGTCTGAATAGCTAACCAAAATCTCCCACAAGATTTATTGCCGTGACCTATGGCCTTTCACGCTGGGCT\n",
      "GATCTGATTCGTAGTAGCTAGTCCCGAAGCTCTTATGAACACGCTGATACGATACAATTGTTGATAAGAGGCCCAACCGCGATAGGATGCGGCTGATATTTGTTATGTCCATGGGCTTCCCCCTTAGGGATCCTAACCTCTGTTCGTATTATTCCG\n",
      "GAAGGTGCCATACCTTTTCCTGGTTCCGACCTTTCGTCTCGATTTAAAGCGTATCTCGCCAGCCGATGGTTCTTGGTCTTAGATAGGCGTGATGCATCTTTCAGACGGGGCCAGCGGCTGGGGAAATTAAGCCAATTGGCTGTCCACGGCGCGATG\n",
      "TCACGTGTTCTTATTGTGTTCCGGCGAAGTTCCGTTTGTCCCTGCGATATGAAGCAGTTATCTCCTAGAATTTCCCGATAAGTCGAAGCTGGTGTTCTATCTTTGCGGCCATTGGCTCTCCAGCATCCGTAAGAGAGACGCGCGAGAAAATCCCAC\n",
      "\n",
      "Best motifs = CAATCGGCTTCC CAATAGGCTCAC CGATGGGCTACC CAATGGGCTCAC CCATGGGCTGCC CAATTGGCTGTC CGATTGGCTCTC CTATTGGCTCGC CTATGGGCTGTC CGATGGGCTGAC CGATGGGCTAAC CCATAGGCTGCC CAATCGGCTTTC CTATTGGCTGCC CCATCGGCTCAC CGATGGGCTCGC CGATGGGCTGGC CCATTGGCTTGC CCATGGGCTCAC CTATCGGCTGTC CTATGGGCTAGC CCATCGGCTCCC CCATGGGCTTCC CAATTGGCTGTC CCATTGGCTCTC\n"
     ]
    }
   ],
   "source": [
    "filename = 'Input_MotifSearch.txt'\n",
    "Genome = []\n",
    "with open('Course1_Data/'+filename,'r',encoding='utf-8') as file:\n",
    "   k,t = tuple([int(i) for i in file.readline().strip().split(' ')])\n",
    "   for i in range(t):\n",
    "       Genome.append(file.readline().strip())\n",
    "    \n",
    "N_iter = 1000\n",
    "BestMotifs = MCMotifSearch(Genome,k,t,N_iter)\n",
    "print(f'k = {k}, t = {t}')\n",
    "print('Genomes = \\n' + '\\n'.join(Genome))\n",
    "print('\\nBest motifs = ' + ' '.join(BestMotifs))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f043020",
   "metadata": {},
   "source": [
    "#### Monte Carlo search with Gibbs Sampling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "5d03fb99",
   "metadata": {},
   "outputs": [],
   "source": [
    "def MCMotifGibbs(DNA, k, t, N=1000, M=10):\n",
    "    \n",
    "    def GibbsSampler(DNA, k, t, N):\n",
    "        from random import randrange\n",
    "        N = len(DNA[0])\n",
    "        BestMotifs = []\n",
    "        for pattern in DNA:\n",
    "            i = randrange(N-k+1)\n",
    "            BestMotifs.append(pattern[i:i+k])\n",
    "        for j in range(N):\n",
    "            i = randrange(t)\n",
    "            motifs = BestMotifs.copy()\n",
    "            motifs.pop(i)\n",
    "            profile = Profile(motifs)\n",
    "            motifi = MostProbable(k, DNA[i], profile)\n",
    "            motifs.insert(i,motifi)\n",
    "            if Score(BestMotifs) > Score(motifs):\n",
    "                BestMotifs = motifs\n",
    "        return Score(BestMotifs), BestMotifs\n",
    "    \n",
    "    BestScore = k*t+1\n",
    "    BestMotifs = None\n",
    "    for i in range(M):\n",
    "        score, motifs = GibbsSampler(DNA, k, t, N)\n",
    "        if BestScore > score:\n",
    "            BestScore = score\n",
    "            BestMotifs = motifs\n",
    "    return BestMotifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "1e7644f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "k = 12, t = 25\n",
      "Genomes = \n",
      "CTATGGGGTATTTTTGTGCGGGTATTAAAGTCGGATAACACCCACCATTACTTGCGTAACCATGGCTGCTAGAGATCTTGTAGCCACGTTTGACCTGTGTGAACTGTACCACGCTCGCCCCAATCGGCTTCCGCCACGCTGTTAAGATCCTACCGG\n",
      "CTGCAGCATAAGGACAACGACTTATCATTTCACCGCCGCCGCGCTACACAATAGGCTCACAGAGCTGACTGCCCTGTCGATTGTCCACAGGCAGGCCCCAATGGGCAAATAGCGAATAGTACCAACCTACGATGTGTTCCGCGGAAGTTACATACA\n",
      "TTGTTTGCCTCCGCTTCTTCGCCCAGTTAGTATTGCTTATAATCCCCGGATGCCATTTCGCAGTGACGATGACGATGGGCTACCCTTGGCCGATTCCTTTCGATGAGCCCTGATGAATCTGCGCGGAATCGGAGGGGGTTAATGGCACGCTAAACT\n",
      "GGTTCATTCGTTGTGTTGAAAAAGGGCACGTCGCCTCAATGGGCTCACGCATAGCTGTAGAGATTCGCGGTACCTGCCTTATAGAGCACCCACCTAAGGTGAGGCCCTATTTCGTAGATTACCTCTACATCAGTACTTATTAACGGGGGAAATACT\n",
      "TTAAGGCCGATAAGTATTTCCCCCGTTAGGTATTCTTACTCTCCTACAATTATCCCTGATGTTACGGTGATTCGTGCAACAACGTGGCAACAGTGTCGAGATGGGAGGCCATGGGCTGCCAGCGCTCATTTGAGTGTTTTCTTGTTCTGTGTCAAT\n",
      "CCGTTCACGTTCATAACAAGCAGGAAAACCTTTGGAGGCTTGGAGTCGGGTTGTTGAATGGTAGCTTAGTATCAATTGGCTGTCTTCTCACTTGCCGGTTTAGTAGACTCAGCTCCGTCTTAGCTCTTTTACGCATCCGAGCCTACCTCTGCGTCT\n",
      "TGCAGTATCGCATGCGCGCACTTCACGATGAAGCTAACGAGCGGACAACATGGTAGGACAGAGAACACTCTCCGTTCAAAATTCCGATTGGCTCTCTTAGGATTGTGCCAGACGGCATGAAAGGGGGATTATTCGATAGAGTGTCCCAACCGACTA\n",
      "ATAGGGACGTATTAGTCATGCTGTCGGACATGGCTTGAACACATTACATAGGGACCGTTGCGTTCAAGAGTTTTACCGTCACCTATGGAATATTTGTGGTGGCGTAAGTCCACCGGGCCAACTGTAAGTTTCACCTGCTAGACGCTATTGGCTCGC\n",
      "GCGGGCTAGGGACTTATTTAAACCACGTTCGTCGGCCGTTCCGTAACTTAACGCCGGAACTGCGGTACTCAACCAATGATACAAGTCCCCTACGCAGCCAAATCACTCGGGATGCGCACAGTCACACCTTGACTATGGGCTGTCCCGACTGCAGTG\n",
      "CGATGGGCTGACTTGCCAGTCTAGGCCAGAATCCCTAGTATTACTGTCGTTATTGTCCTCGATACGCACCTAACATTACTACGTATGGGTGTCTTGGCTTAGGGCGACCTATGCTCAGAATTAGAGATGCTCAACTGCAACTTCTGCAACGAAGTG\n",
      "GAGTCATGACGGGCTCTTGAGGCCCCGATGCGGCTGGGTAACAAGACGCGCAATGTTAGCACTACTTACCATAGTCTAGACGTGAACATGAATAGGGTGACAGCTGATGTAGACAACAGATGCCACGGATGCACTCCTCGTTAGCGATGGGCTAAC\n",
      "ATGAATTCTTTTAAAGTTTGTTCACGCTACTCTGCGTCTCCTAGGTCTCCATAGGCTGCCTCCCCTTAGGGAGGTAGAAGAACATAAATTACCTCGAATCCGGCATCTCGTTTGAGTCGTACCGTCATCAGAGGAAAGTGTCGGGATAAAACGGGA\n",
      "AGCGGAGGGTGATTGGACGCCTTAGAGATGCAGAGTAAGTGTGATTATCTAAATTAACAGTATTTGACAGTGTCACACGCTACAGTTTCCCGAATCTAAAAGACATATCAATCGGCTTTCACTCAGTCACCTTTAAGTACCAACGCCCGGTCGTAA\n",
      "GTGACTAATATCAGAGGTATAGAGATTCATTGGTTGAAACGAGCCGTTCTATTGGCTGCCGGAGAATTAATCGTGGACCATACATTGGTGCCACCCGCGGTCGGCTTGTGGTTTCATTGGAGGGCGACTGTGCAATACCACAGCACACCAGTGTCT\n",
      "TGACCCGTTACACATTTGCAGCTGGTTTTTTCCACATGTGTCATGCTCCAGCGGTTGATCACTGCAGGGAACACTAGCGAAAAAGTGCTTACGAGTCAACTTTCATATATCACACAGACGGCCAGGACATTTCCATCGGCTCACCCGGTGGGATTA\n",
      "CGCTCTCTGGTTGTGTAATGGAACGAAGCCCCCATGGGTTGCAGGAGTCGGCGACACCAACAAACAGCATACCAACGGCCAAGTATGCTCTACATCCGATGGGCTCGCCCTGGACCCGATGTAATTACTCGATCGGAACCCTTATACCCAGCGTGT\n",
      "GACACCTGGCCTGCCGAGGAAGAGAGAAACACGATCTCACCCCACTTCCGTTAATAGAAAGGTGTATGCTTGGACCGGTAACCTCGGCTTAGTTTGGGCTGTCGTCATCGATGGGCTGGCCGGAATCGAGGGAGTAGCCATGTCGCTGTCTCGATC\n",
      "GGACAGTCATTGAGCTATTGGGTACGCCTTCGAGTTTGATGGTCTCTCCTCGAGCCCTGTCAACTGAAGTAGTAGTTATGGCCAGCCCCAGCTGCCCTGCCATACACATCGCCGCGATTGATGTCCATTTTTCCATTGGCTTGCTCTGAAAAGTAA\n",
      "ACGTGGCAGAAGGCCTCGATCAAGTTCAAGAGTCTAATTGGTTCGTGGGTTGTTACAACAATTATTTTCTGACCATGGGCTCACAGACGCGCACATGGACGATAACGGCGGGCATAGGTCTTCTTGTCGAAGGAAGCCTCCCATGGGAAGCGACCG\n",
      "CTATAAATACGCCTCGAACATGCGGCACCAGTTTTCTCGGATTCTGGCCTGGAGGTGAGTTGTACCATCAAGATGTAGTGCTACCTATCGGCTGTCTGGTCTGAAATTACCCTTTAGATCCCAATAGTGCGGGAATACCGCTTTGAATTTACTCAT\n",
      "CGCGATCATCTCGAGCTACTTATATATACTCTTGTAGCCGATTGTCCTCTATGGGCTAGCTTCTATACTGAGTCAGTATCCGCCCCGTCAGGATCTATCCAGGTAGGCGACCTTACAACCACTGGCAAGGTAACCTGTTGTGGGCTCGTAGCGCAA\n",
      "TCATGGGTCGTTTGGCACCTGCAGACAATAGTACAGCTCGTACCACTTGCAGAGGTAAATTCCATCGGCAGTCCATCGGCTCCCTGATGTCTGAATAGCTAACCAAAATCTCCCACAAGATTTATTGCCGTGACCTATGGCCTTTCACGCTGGGCT\n",
      "GATCTGATTCGTAGTAGCTAGTCCCGAAGCTCTTATGAACACGCTGATACGATACAATTGTTGATAAGAGGCCCAACCGCGATAGGATGCGGCTGATATTTGTTATGTCCATGGGCTTCCCCCTTAGGGATCCTAACCTCTGTTCGTATTATTCCG\n",
      "GAAGGTGCCATACCTTTTCCTGGTTCCGACCTTTCGTCTCGATTTAAAGCGTATCTCGCCAGCCGATGGTTCTTGGTCTTAGATAGGCGTGATGCATCTTTCAGACGGGGCCAGCGGCTGGGGAAATTAAGCCAATTGGCTGTCCACGGCGCGATG\n",
      "TCACGTGTTCTTATTGTGTTCCGGCGAAGTTCCGTTTGTCCCTGCGATATGAAGCAGTTATCTCCTAGAATTTCCCGATAAGTCGAAGCTGGTGTTCTATCTTTGCGGCCATTGGCTCTCCAGCATCCGTAAGAGAGACGCGCGAGAAAATCCCAC\n",
      "\n",
      "Best motifs = AATCGGCTTCCG AATAGGCTCACA GATGGGCTACCC AATGGGCTCACG CATGGGCTGCCA AATTGGCTGTCT GATTGGCTCTCT CATGGCTTGAAC TATGGGCTGTCC GATGGGCTGACT AATAGGGTGACA CATAGGCTGCCT AATCGGCTTTCA TATTGGCTGCCG CATCGGCTCACC GATGGGCTCGCC GATGGGCTGGCC CATTGGCTTGCT CATGGGCTCACA TATCGGCTGTCT TATGGGCTAGCT CATCGGCAGTCC CATGGGCTTCCC AATTGGCTGTCC CATTGGCTCTCC\n"
     ]
    }
   ],
   "source": [
    "filename = 'Input_MotifSearch.txt'\n",
    "Genome = []\n",
    "with open('Course1_Data/'+filename,'r',encoding='utf-8') as file:\n",
    "   k,t = tuple([int(i) for i in file.readline().strip().split(' ')])\n",
    "   for i in range(t):\n",
    "       Genome.append(file.readline().strip())\n",
    "    \n",
    "N_repeat = 1000\n",
    "BestMotifs = MCMotifGibbs(Genome,k,t)\n",
    "print(f'k = {k}, t = {t}')\n",
    "print('Genomes = \\n' + '\\n'.join(Genome))\n",
    "print('\\nBest motifs = ' + ' '.join(BestMotifs))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
